File: Python/239. Sliding Window Maximum.py
# 用单调队列来从大到小存放滑动窗口的值
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        class MyQue:
            def __init__(self):
                self.que = []

            def push(self, val):
                while self.que and self.que[-1] < val:
                    self.que.pop(-1)
                self.que.append(val)

            def pop(self, val):
                if self.que and self.que[0] == val:
                    self.que.pop(0)

            def empty(self):
                return len(self.que) == 0

            def front(self):
                return self.que[0]

        que = MyQue()
        res = []
        for i in range(k):
            que.push(nums[i])
        res.append(que.front())
        for i in range(k, len(nums)):
            que.pop(nums[i - k])
            que.push(nums[i])
            res.append(que.front())
        return res

File: C++/239. Sliding Window Maximum.cpp
//用单调队列来从大到小存放滑动窗口的值 
class Solution {
public:
    class MyQueue {
    public:
        deque<int> que;
        void push(int val) {
            while (!que.empty() && que.back() < val) que.pop_back();
            que.push_back(val);
        }

        void pop(int val) {
            if (!que.empty() && que.front() == val) que.pop_front();
        }

        int front() {
            return que.front();
        }
        bool empty() {
            return que.empty();
        }
    };
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        MyQueue que;
        for (int i = 0; i < k; i++) {
            que.push(nums[i]);
        }
        vector<int> vec;
        vec.push_back(que.front());
        for (int i = k; i < nums.size(); i++) {
            que.pop(nums[i - k]);
            que.push(nums[i]);
            vec.push_back(que.front());
        }
        return vec;
    }
};